sc_demographics.R

```diff
@@ -0,0 +1,188 @@
+# Description -
+#####################################################
+
+
+## output ##
+output <- haven::read_sav(file = "/conf/hscdiip/SLF_Extracts/Social_care/sc_demographics_lookup_Dec_2021.zsav")
+
+
+
+## load packages ##
+library(readr)
+library(odbc)
+library(dplyr)
+library(stringr)
+library(dbplyr)
+library(phsmethods)
+
+
+
+## data ##
+
+######################################################
+# set-up conection to platform
+db_connection <- odbc::dbConnect(
+  odbc::odbc(),
+  dsn = "DVPROD",
+  uid = Sys.getenv("USER"),
+  pwd = rstudioapi::askForPassword("password")
+)
+###################################################
+
+# read in data - social care 2 demographic
+sc_demog <- tbl(db_connection, in_schema("social_care_2", "demographic")) %>%
+  select(
+    latest_record_flag, extract_date, sending_location, social_care_id, upi,
+    chi_upi, submitted_postcode, chi_postcode, submitted_date_of_birth,
+    chi_date_of_birth, submitted_gender, chi_gender_code
+  ) %>%
+  collect()
+
+# variable types
+sc_demog <- sc_demog %>%
+  mutate(
+    submitted_gender = as.numeric(submitted_gender),
+    chi_gender_code = as.numeric(chi_gender_code)
+  )
+
+
+## clean up data ##
+sc_demog <- sc_demog %>%
+  mutate(
+    # use chi if upi is NA
+    upi = coalesce(upi, chi_upi),
+    # check gender code - replace code 99 with 9
+    submitted_gender = replace(submitted_gender, submitted_gender == 99, 9)
+  ) %>%
+  mutate(
+    # use chi gender if avaliable
+    gender = if_else(is.na(chi_gender_code) | chi_gender_code == 9, submitted_gender, chi_gender_code),
+    # use chi dob if avaliable
+    dob = coalesce(chi_date_of_birth, submitted_date_of_birth)
+  )
+
+
+## postcode ##
+# clean-up
+sc_demog <- sc_demog %>%
+  mutate(
+    # remove any postcodes with length < 5
+    submitted_postcode = na_if(submitted_postcode, str_length(submitted_postcode) < 5 | str_length(submitted_postcode) == 5),
+    chi_postcode = na_if(chi_postcode, str_length(chi_postcode) < 5 | str_length(chi_postcode) == 5)
+  ) %>%
+  mutate(
+    # remove spaces
+    submitted_postcode = str_replace_all(submitted_postcode, fixed(" "), ""),
+    chi_postcode = str_replace_all(chi_postcode, fixed(" "), "")
+  ) %>%
+  mutate(
+    # remove postcodes with length 8 or longer
+    submitted_postcode = na_if(submitted_postcode, str_length(submitted_postcode) > 8 | str_length(submitted_postcode) == 8),
+    chi_postcode = na_if(chi_postcode, str_length(chi_postcode) > 8 | str_length(chi_postcode) == 8)
+  )
+
+
+# format postcodes using `phsmethods`
+sc_demog <-
+  sc_demog %>%
+  mutate(across(contains("postcode"), ~postcode(.x, format = "pc7")))
+
+
+# count number of na postcodes
+na_postcodes <-
+  sc_demog %>%
+  count(across(contains("postcode"), ~is.na(.x)))
+
+# UK postcode regex
+uk_pc_regexp <- "([Gg][Ii][Rr] 0[Aa]{2})|((([A-Za-z][0-9]{1,2})|(([A-Za-z][A-Ha-hJ-Yj-y][0-9]{1,2})|(([A-Za-z][0-9][A-Za-z])|([A-Za-z][A-Ha-hJ-Yj-y][0-9][A-Za-z]?))))\\s?[0-9][A-Za-z]{2})"
```

Comment:
Simplify for readability (and also to avoid false negatives (after reading this: [https://ideal-postcodes.co.uk/guides/postcode-validation](https://ideal-postcodes.co.uk/guides/postcode-validation))
```suggestion
# UK postcode regex - see https://ideal-postcodes.co.uk/guides/postcode-validation
uk_pc_regexp <- "^[a-z]{1,2}\\d[a-z\\d]?\\s*\\d[a-z]{2}$"
```

This will likely exclude fewer postcodes than before, but at least we can be certain we're not excluding a potentially valid postcode by accident.